mimic paradise/16159

Spookerton opened the pull request:
adds SSping, moves chat ping behavior to it instead of sleeping
adds convenience define RUNLEVELS_ALL
cleans up some style & loop bits since we're in there

Spookerton committed:
mimic paradisess13/paradise/pull/16159

Spookerton committed:
style pass on chat & output, tiny tweaks, globals

chatOutput instantiates lists outside New
SSchat payloads is static
SSchat loop in fire made insubstantially skinnier
CLIENT_FROM_VAR -> resolve_client

SierraKomodo commented on code/controllers/subsystems/chat.dm:
```diff
@@ -4,75 +4,69 @@ SUBSYSTEM_DEF(chat)
 				continue
-
 			payload[C] += twiceEncoded
-
 	else
-		var/client/C = CLIENT_FROM_VAR(target) //Grab us a client if possible
-
-		if(!C)
+		var/client/C = resolve_client(target)
+		if (!C)
 			return
-
-		//Send it to the old style output window.
-		legacy_chat(C, original_message)
-
-		if(!C?.chatOutput || C.chatOutput.broken) //A player who hasn't updated his skin file.
+		legacy_chat(C, original_message) //Send it to the old style output window.
+		if (!C?.chatOutput || C.chatOutput.broken) //A player who hasn't updated his skin file.
 			return
-
-		if(!C.chatOutput.loaded) //Client still loading, put their messages in a queue
+		if (!C.chatOutput.loaded) //Client still loading, put their messages in a queue
 			C.chatOutput.messageQueue += message
 			return
-
 		payload[C] += twiceEncoded
```
While we're here, all of this should really be moved to a separate proc with proc calls to reduce the copy+paste. (Not a requirement for approval)

Spookerton replied:
I'm inclined to say I'd rather keep things unrolled since every dm land call adds overhead and chat is `somewhat important`, but it can't be all that bad I suppose

SierraKomodo commented on code/modules/goonchat/browserOutput.dm:
```diff
@@ -2,97 +2,82 @@
-	var/cookieSent   = FALSE // Has the client sent a cookie for analysis
-	var/broken       = FALSE
-	var/list/connectionHistory //Contains the connection history passed from chat cookie
+	var/client/owner //! client ref
+	var/total_checks = 0 //! How many times client data has been checked
+	var/next_time_to_clear = 0 //! When to next clear the client data checks counter
+	var/loaded = FALSE //! Has the client loaded the browser output area?
+	var/list/messageQueue = list() //! If they haven't loaded chat, this is where messages will go until they do
+	var/cookieSent = FALSE //!  Has the client sent a cookie for analysis
+	var/broken = FALSE //! The client's skin prevented loading
+	var/list/connectionHistory = list() //! Contains the connection history passed from chat cookie
```
Same-line dmdocs do not parse correctly in VSC, although the `//!` prefix does not appear to be parsed at all as a dmdoc.

Spookerton replied:
A great sadness, that <https://github.com/SpaceManiac/SpacemanDMM/tree/master/src/dmdoc#documenting-code> lists them, then

SierraKomodo commented on code/modules/goonchat/browserOutput.dm:
```diff
@@ -2,97 +2,82 @@
-	// Build arguments.
-	// Arguments are in the form "param[paramname]=thing"
-	var/list/params = list()
+	var/list/params = list() //! Build proc parameters from the form "param[paramname]=thing"
```
As above

Spookerton committed:
chatOutput style 2

//! same-line to /// prior-line
stripped datum prefix
collected def and inst of client/chatOutput to chatOutput file
queue code dupe -> queuePartTwo

+ globals hash to pacify tests

SierraKomodo reviewed (dismissed):
Test fails - Globals stuff. Not seeing anything else that should need changing.

SierraKomodo reviewed (approved):
My gut tells me this is either going to explode horribly or explode wonderfully. Full send.
